src/internet_identity/src/main.rs

```diff
@@ -244,6 +303,211 @@ async fn init_salt() {
+        trap_if_not_authenticated(entries.iter().map(|e| &e.pubkey));
+
+        prune_expired_tentative_device_registrations(state);
+
+        state
+            .tentative_device_registrations
+            .borrow_mut()
+            .remove(&user_number);
+    })
+}
+
+#[update]
+async fn add_tentative_device(
+    user_number: UserNumber,
+    device_data: DeviceData,
+) -> AddTentativeDeviceResponse {
+    let verification_code = generate_device_verification_code().await;
+
+    match check_tentative_device_registration_prerequisites(user_number) {
+        Ok(device_registration_timeout) => {
+            STATE.with(|state| {
+                state.tentative_device_registrations.borrow_mut().insert(
+                    user_number,
+                    DeviceTentativelyAdded {
+                        expiration: device_registration_timeout,
+                        tentative_device: device_data,
+                        verification_code: verification_code.clone(),
+                        failed_attempts: 0,
+                    },
+                )
+            });
+            AddTentativeDeviceResponse::AddedTentatively {
+                verification_code,
+                device_registration_timeout,
+            }
+        }
+        Err(err) => err,
+    }
+}
+
+#[update]
+async fn verify_tentative_device(
+    user_number: UserNumber,
+    user_verification_code: DeviceVerificationCode,
+) -> VerifyTentativeDeviceResponse {
+    match check_tentative_device_code_counting_failed_attempts(user_number, user_verification_code) {
+        Ok(device) => {
+            add(user_number, device).await;
+            VerifyTentativeDeviceResponse::Verified
+        }
+        Err(err) => err,
+    }
+}
+
+/// Checks the device verification pin for a tentative device.
+/// If valid, returns the device to be added and exits device registration mode
+/// If invalid, returns the appropriate error to send to the client and increases failed attempts. Exits device registration mode if there are no retries left.
+fn check_tentative_device_code_counting_failed_attempts(
+    user_number: UserNumber,
+    user_verification_code: DeviceVerificationCode,
+) -> Result<DeviceData, VerifyTentativeDeviceResponse> {
+    STATE.with(|state| {
+        let entries = state
+            .storage
+            .borrow()
+            .read(user_number)
+            .unwrap_or_else(|err| {
+                trap(&format!(
+                    "failed to read device data of user {}: {}",
+                    user_number, err
+                ))
+            });
+        trap_if_not_authenticated(entries.iter().map(|e| &e.pubkey));
+
+        prune_expired_tentative_device_registrations(state);
+
+        let mut device_registration_state = state.tentative_device_registrations.borrow_mut();
+        match device_registration_state.remove(&user_number) {
+            None | Some(DeviceRegistrationModeActive { .. }) => Err(WrongCode { retries_left: 0 }),
```

Comment:
Was looking at the backend tests and got surprised by this. Do we reply with "wrong code" for security reasons? If so can you add a note?

Reply:
Yes, we discussed this earlier. But actually, I just realized that it is useless as this is an authenticated call. I'll revert to giving descriptive errors.